satpy/readers/utils.py

```diff
@@ -474,3 +475,53 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
     with xr.set_options(keep_attrs=True):
         reflectance = reflectance / reflectance.dtype.type(sun_earth_dist * sun_earth_dist)
     return reflectance
+
+
+def get_distributed_friendly_dask_array(manager, varname, chunks, dtype,
+                                        group="/", auto_maskandscale=None):
+    """Construct a dask array from a variable for dask distributed.
+
+    When we construct a dask array using da.array and use that to create an
+    xarray dataarray, the result is not serialisable and dask graphs using
+    this dataarray cannot be computed when the dask distributed scheduler
+    is in use.  To circumvent this problem, xarray provides the
+    CachingFileManager.  See GH#2815 for more information.
+
+    Should have at least one dimension.
+
+    Example::
+
+        >>> import NetCDF4
+        >>> from xarray.backends import CachingFileManager
+        >>> cfm = CachingFileManager(NetCDF4.Dataset, fn, mode="r")
+        >>> arr = get_distributed_friendly_dask_array(cfm, "my_var")
+
+    Args:
+        manager (xarray.backends.CachingFileManager):
+            Instance of xarray.backends.CachingFileManager encapsulating the
+            dataset to be read.
+        varname (str):
+            Name of the variable.
+        chunks (tuple):
+            Chunks to use when creating the dask array.
+        dtype (dtype):
+            What dtype to use.
+        group (str):
+            What group to read the variable from.
+        auto_maskandscale (bool, optional):
+            Apply automatic masking and scaling.  This will only
+            work if CachingFileManager.acquire returns a handler with a
+            method set_auto_maskandscale, such as is the case for
+            NetCDF4.Dataset.
+    """
+    def get_chunk():
+        with manager.acquire_context() as nc:
+            if auto_maskandscale is not None:
+                nc.set_auto_maskandscale(auto_maskandscale)
+            return nc["/".join([group, varname])][:]
```

Comment:
This code is specific to `netCDF4`. While I understand that is the use case here, it suggests that either the function name and purpose needs to change or the `set_auto_maskandscale` functionality should be changed. Along with that, if I'm understanding the CFM properly, the `nc` object here should be the same between accesses. I mean, the exact same object if accessed in the same process. Although unlikely in the current usage setting this global attribute on the `netCDF4.Dataset` object could cause race conditions or unexpected behavior...although very unlikely.

To resolve both of these concerns, could this function and the current usage be changed to:

1. Use a wrapper (or `functools.partial` is possible) around `netCDF4.Dataset` when creating the CFM to set the auto_maskandscale option when `netCDF4.Dataset` is called inside the CFM. Although more hacky, there could also be a callback/init_func or maybe getitem function (see next 2 points) that handles setting of these per-variable things there (auto mask and scale can be set on the file object or the variable object).
2. The `group` and `varname` being joined is NetCDF specific and as far as I can tell isn't needed if the user is just expected to do it themselves.
3. This function assumes that `__getitem__` works for whatever file-like object is accessed and that `[:]` is needed. I don't think `__getitem__/[]` access for a variable is a bad assumption, but if there is a way to make this generic maybe that's what we need. If this function ended up being used with another library for reading files then the `[:]` may make an unnecessary copy and may not be noticeable to someone using the function.

Reply:
I've added a wrapper along with `functools.partial` and moved the `set_auto_maskandscale` there.  I've only set this globally; I find per variable too complicated and we set it globally so far already, so there is no change there.  I've also changed the group handling and moved this responsibility to the user.

As for the `__getitem__` assumption, yes, there is such an assumption.  I'm not sure how to make this generic without making it harder to use.  To make it generic, the user would need to pass information in some way on how the variable is obtained, via a callable or a wrapper class.  Do you have an idea on what would be a good way to implement this?

Reply:
I'm ok with the `__getitem__` assumption and even the `[:]` if it just seems annoying to make it more generic. One option could be to do something like `if not isinstance(x, np.ndarray)` (or some other "is not numpy-like or is netcdf Variable-like object" type of check) then you could do the `[:]` inside that `if` block.

Reply:
I replaced `[:]` by taking information from `block_info`:

```
    def get_chunk(block_info=None):
        arrloc = block_info[None]["array-location"]
        with manager.acquire_context() as nc:
            var = nc[varname]
            return var[tuple(slice(*x) for x in arrloc)]
```

It's less obvious than `[:]` (at least to me) but I think it's correct :D

If something doesn't support such indexing, then it will fail with an exception and the user will notice that it doesn't work.

Reply:
Sounds good.